src/YoutubeAtomPlayer.tsx

```diff
@@ -273,16 +284,14 @@ export const YoutubeAtomPlayer = ({
                     }
                 };
 
-                player.current && player.current.on('ready', readyListener);
+                const readyListener = player.current?.on(
```

Comment:
@arelra I noticed in the docs for [youtube-player](https://github.com/gajus/youtube-player) that they create a variable for each event listener which is then used as the argument for `player.off`. Using this approach here seems to resolve the issue with crashing.

```
var listener = player.on(/* ... */);

player.off(listener);
```

Reply:
Although this works, an issue now is that when the `videoControls` prop is changed, the cleanup runs and the listeners are not reinitiated, so the tracking events are no longer fired.

Reply:
> I noticed in the docs for [youtube-player](https://github.com/gajus/youtube-player) that they create a variable for each event listener

Good catch!

Reply:
> Is there a way we can reset the listeners?

I think the way the listeners currently get unregistered is incorrect and we can improve it:

#### Problem(s)

I think there's a few different issues combined here!

1. Another subtle useEffect foot-gun to add to the collection, is that useEffect cleanup functions run on _every_ useEffect trigger i.e. if a dependency has changed:
	- https://maxrozen.com/demystifying-useeffect-cleanup-function

1. Storybook components are not passing props to `YoutubeAtom` as `YoutubeBlockComponent` does:
    
    - In Storybok the `eventEmitters` prop is created for each render which causes the `initPlayer` useEffect to trigger for every re-render in Storybook.
    - When used via `YoutubeBlockComponent`, `eventEmitters` are 'stable' so the cleanup function does not trigger on re-renders.
    
1. New re-renders introduced :
	
	- Previously `YoutubeAtomPlayer` would only render once so we never saw any re-render related issues.
	- Setting `videoControls` causes x2 re-renders (first to set the control, second to reset it to undefined)
	- This is not a problem per se but in Storybook re-renders are causing the useEffect cleanup to run which is pointing to a problem with the useEffect setup.

1. Incorrect placement of listener cleanup:

	- Because of 1) and 3) I think the current placement of the listener cleanup function (carried over from ye olde component) in the `initPlayer` useEffect is incorrect as the cleanup will run on every change to one of its dependencies.
	- In practise this isn't a problem as none of its dependencies _should_ change but in Storybook they do because of the unstable `eventEmitters`.
		
#### Fix

I _think_ a 'proper' fix is to separate the listener cleanup from the `initPlayer` useEffect to ensure it only runs on component unmount.

To do this I've:

- stored the listeners on a ref
- added a separate useEffect to remove the listeners with an empty dependencies array. This should only remove the listeners on unmount.

It should now work with no console errors and event emitters should be unaffected.

I've not tested this exhaustively so please test and verify :)

Reply:
This great, thank you! I've been doing some testing and it all works nicely 👌